### ISSUE TITLE
jobattach() function for adding handlers to an existing job

### ISSUE THREAD

**Issue by justinmk** [neovim/neovim#5023]
AFAIK this doesn't exist: ability to attach `on_stdout`/`on_stderr`/etc handler to an existing job. Something like this:

    call jobattach({jobid}, {opts})

where `{opts}` is a dictionary whose structure is the same as that of `jobstart()`.

In fact I think we should deprecate `jobstart()`, rename it `job()`, and if the first argument of `job()` is a job id, operate on the existing job.

**Comment by bfredl** [comment]
> ability to attach on_stdout/on_stderr/etc handler to an existing job

Is there an intended use case for this? What happens if the job already has an handler of this kind, does it get replaced, or will both fire?

> In fact I think we should deprecate jobstart(), rename it job()

Why? jobstart is a good descriptive name, why obfuscate it?

**Comment by justinmk** [comment]
> use case for this?

Use case is listening to `:term` output: http://stackoverflow.com/q/38216231/152142

> What happens if the job already has an handler of this kind, does it get replaced, or will both fire?

It calls all attached handlers.

> Why? jobstart is a good descriptive name, why obfuscate it?

I won't insist on it. Maybe I am veering into too-clever territory.

**Comment by bfredl** [comment]
> Use case is listening to :term output

isn't `:term` is just thin wrapper around `termopen()`, which already would allow this? Could be a naming/documentation issue.

**Comment by justinmk** [comment]
Oh god, I forgot about `termopen()`. 

But then some questions:

-  why do we expose `b:terminal_job_id` _at all_ if we insist on `termopen()` for programmatic control? 
- is there much reason _not_ to allow attaching additional handlers to a job?

**Comment by bfredl** [comment]
> why do we expose b:terminal_job_id at all

Why not? otherwise _any_ repl managing plugin would be forced to reimplement this themselves. (_regardless_ if it only manages terminals it opens itself with `termopen` or it provides functionallity to "unowned" terminals opened by the user)

> is there much reason not to allow attaching additional handlers to a job?

Not really, except for making the code more complex, but it shouldn't be too bad (keep a list of self-ufunc_t pairs instead of one on each). I was just curious if this actually was a requested feature.

**Comment by bfredl** [comment]
There is also the slight issue of racyness, if someone does
```
let job = jobstart()
call job_attach(job, {'on_stdout':...})
```
do we promise it to be race-free? Or slightly more relevantly
```
let job = jobstart()
call do_something()
call job_attach(job, {'on_stdout':...})
```
where it is not entirely obvious to what extent `do_something()` invokes event processing.

**Comment by justinmk** [comment]
> Why not? otherwise any repl managing plugin would be forced to reimplement this themselves. (regardless if it only manages terminals it opens itself with termopen or it provides functionallity to "unowned" terminals opened by the user)

I'd say the same use case of "providing functionality to unowned terminals" drives the request for `jobattach()`: if a plugin needs to listen to a `:term` opened by the user or some other plugin.

> do we promise it to be race-free? 

Do you mean because a `on_stdout` must be evaluated, and can have side-effects? I think it should be treated (in documentation and implementation) the same as autocmd handlers: if multiple `autocmd` handlers are listening to the same event, then:

- the first such handler is processed, its side-effects are reconciled, 
- then the next handler is processed, (sending the same data as was sent to the first handler)
- etc...

**Comment by bfredl** [comment]
No I meant: could some prefix of the output already be processed before the handler gets registered and thus not seen by the handler? Reasonably not in the first case, but consider if `do_something` does "unrelated" event processing.

**Comment by justinmk** [comment]
> could some prefix of the output already be processed before the handler gets registered and thus not seen by the handler

Not sure I see the problem. Any handler that's added after the first one must expect that it could miss data since the initial call. We can document that to make it extra clear.

**Comment by bfredl** [comment]
Right, it is a problem of documentation and expectation management. But if we include say a streamreader helper for plugins that only are interested in the complete stdout on exit, the question is if we want to allow the interface
```
    let job = jobstart(...)
    call util#readstream(job, "stdout", "on_complete_output")
```
where "on_complete_output" is called on exit/stdout-eof with the complete output (with partial lines properly joined)
or if we are prefer an interface that don't need this:
```
  let job = jobstart(..., {"on_stdout": "util#streamreader", 
                          \ "on_complete_stdout":"callback"})
```
or even
```
call util#systemlist_async(..., "exit_callback", stdin_data)
```
where the exit callback will receive the complete outputs.
Not saying the first option is the best, just that it is an opportunity to consider.

**Comment by justinmk** [comment]
Instead of `on_complete_stdout`, could we send the entire data as an argument to `on_exit`?

**Comment by bfredl** [comment]
> This could be enabled optionally as a flag in the {opt} dictionary

This usecase seems common enough to warrant a simple convenient solution, so sounds reasonable to me.
